arch/x86: Initial Intel64 support

gnuless opened the pull request:
Modify x86 arch to enable 64-bit (LP64) mode with `CONFIG_X86_LONGMODE=y` on boards which support it (for now, Apollo Lake boards).  Added `qemu_x86_long` board as a target for `sanitycheck`. Spot-tested on real hardware (`up_squared`).

Commit:
arch/x86: rename X64 references to Intel64

Intel's X86-64 implementation is officially "Intel64", so follow suit.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
soc/x86/apollo_lake: architecture is goldmont

Update the -march flag passed to the compiler to reflect the
underlying microarchitecture is 'goldmont'.

Fixes: #17104

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
cmake: select appropriate Zephyr SDK compiler for CONFIG_X86_LONGMODE

We need to use the new x86_64 chain provided by the SDK in long mode.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86/pcie: use Z_IRQ_TO_INTERRUPT_VECTOR() macro

The _irq_to_interrupt_vector[] array shouldn't be accessed directly,
as there is a macro for this.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: refactor offsets.c

The IA32 and Intel64 subarchitectures will generate different offset
symbols, so they are refactored. No functional change.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: refactor arch.h

Some of the elements of arch.h can be shared between subarches,
so put them in a common file and factor out the rest. Placeholder
left for the Intel64 definitions to be added later.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: add register definitions to msr.h

Add definitions for the EFER MSR.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: refactor kernel_arch_data.h

Some definitions may be shared between subarchitectures, so refactor
accordingly. The definitions are also modified to separate bits. A
placeholder is created for the Intel64 definitions.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: add placeholders for Intel64 headers

Use different headers for kernel_arch_{func,thread}.h when in
CONFIG_X86_LONGMODE, and add placeholders for Intel64 versions.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: initial Intel64 bootstrap framework

This patch adds basic build infrastructure, definitions, a linker
script, etc. to use the Zephyr and 0.10.1 SDK to build a 64-bit
ELF binary suitable for use with GRUB to minimally bootstrap an
Apollo Lake (e.g., UpSquared) board. The resulting binary can hardly
be called a Zephyr kernel as it is lacking most of the glue logic,
but it is a starting point to flesh those out in the x86 tree.

The "kernel" builds with a few harmless warnings, both with GCC from
the Zephyr SDK and with ICC (which is currently being worked on in
a separate branch). These warnings are either related to pointer size
differences (since this is an LP64 build) and/or dummy functions
that will be replaced with working versions shortly.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: share implementations of sys_io primitives

The sys_inXX() and sys_read/writeXX() I/O primitives can be shared
between subarches of x86 with minor modifications, so move them.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: fix multiboot.c pointer cast

Widen the integer to pointer size before conversion, to make
explicit the intent (and silence the compiler warning). Also
fix a minor bug involving a duplicate (and thus dead) store.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: initial Intel64 support

First "complete" version of Intel64 support for x86. Compilation of
apps for supported boards (read: up_squared) with CONFIG_X86_LONGMODE=y
is now working. Booting, device drivers, interrupts, scheduling, etc.
appear to be functioning properly. Beware that this is ALHPA quality,
not ready for production use, but the port has advanced far enough that
it's time to start working through the test suite and samples, fleshing
out any missing features, and squashing bugs.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: rework z_arch_irq_lock() and z_arch_irq_unlock()

Simplified these two inline functions somewhat, and refactored them:
unlocking can be shared between subarches, but locking cannot.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: set default stack sizes

Using the arch Kconfig here, instead of kernel/Kconfig. Intel64 with
the SysV ABI requires some pretty big stacks. These 4K-8K defaults
are arguably a bit small, but the Zephyr defaults are REALLY too small.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: add IRQ offloading to Intel64 subarch

The IRQ_OFFLOAD_VECTOR config option is also moved to the arch level,
as it is shared between both 32- and 64-bit subarches.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: Add exception reporting code for Intel64

Fleshed out z_arch_esf_t and added code to build this frame when
exceptions occur. Created a separate small stack for exceptions and
shifted the initialization code to use this instead of the IRQ stack.

Moved IRQ stack(s) to irq.c.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: add nested interrupt support to Intel64

Add support for multiple IRQ stacks and interrupt nesting.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: complete 64-bit linker script

The previous linker script was barebones and non-standard. It is
replaced with a script conforms to the rest of the Zephyr arches,
utilizing include/linker headers and standard macros.

link-tool-gcc.h is updated to account for the "i386:x86-64" arch and
the generation of 64-bit ELF binaries.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
boards/x86: add Intel64 (64-bit long-mode) QEMU target board

Add qemu_x86_long board (with CONFIG_X86_LONGMODE=y) for testing.

This requires adding support to soc/ia32 for long mode (trivial),
and adding a quick 64- to 32-bit ELF conversion before invoking
QEMU, which apparently doesn't like 64-bit kernel files.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: refactor z_do_read_cpu_timestamp32()

This function, like its cousin z_tsc_read(), can be shared between
subarches. Note: it's unclear to me if we actually need these to be
two separate functions (do we need serialization?) or if they can
be combined into one function. Add that to the to-do list.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

andrewboie commented on cmake/emu/qemu.cmake:
```diff
@@ -224,6 +224,17 @@ if(CONFIG_X86_IAMCU)
 endif()
 
+if(CONFIG_X86_LONGMODE)
+  # QEMU doesn't like 64-bit ELF files. Since we don't use any >4GB
```
Huh, interesting. I wonder if upstream is fixing this.

gnuless replied:
I don't think they plan to.  I think the concern is that -- in the absence of a more developed firmware environment like UEFI -- simply transferring control to a 64-bit entry point doesn't make sense. I think they're trying to avoid someone slapping together a 64-bit ELF binary and expecting it to work (and then complaining about it).

tyrantlizard77 reviewed (commented):
still looking this over, but so far seems like this is the right approach. appreciate that we are adding a new build target to ensure this doesn't bit-rot.

Commit:
arch/x86: select CONFIG_64BIT when CONFIG_X86_LONGMODE

We need a 64-bit clean kernel when in long mode.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: refactor bitfield-manipulation functions

These inline-assembly functions are identical between IA32 and
Intel64 arches, so move them to the common arch.h file.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: add Z_ARCH_THREAD_STACK_LEN() macro

Implement standard Z_ARCH_THREAD_STACK_LEN() arch interface macro.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
lib/libc/newlib: adjust malloc hooks for Intel64

Intel64 uses the entirety of the "SRAM" device as a user heap.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
subsys/cpp: use POINTER_TO_UINT() when processing constructors

On systems with 64-bit pointers, the direct cast to unsigned
int draws a compiler warning because of mismatched sizes.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: disable inapplicable tests for Intel64

In long mode, x86 does not support static IDTs or OpenOCD,
so disable the tests related to these features.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: tests/cmsis: disable some CMSIS tests in 64-bit mode

The CMSIS layer doesn't like large stacks, and x86 in long mode
requires large stacks, so we disable these tests for now (as was
previously done on the x86_64 port for the same reasons).

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
tests/posix/common: make 64-bit clean

Use the INT_TO_POINTER() and companion macros to silence compiler
warnings about casting between pointers and ints of different sizes.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: add SSE floating-point to Intel64 subarch

This is a naive implementation which does "eager" context switching
for floating-point context, which, of course, introduces performance
concerns. Other approaches have security concerns, SMP implications,
and impact the x86 arch and Zephyr project as a whole. Discussion is
needed, so punting with the straightforward solution for now.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: add support for CONFIG_STACK_SENTINEL

Apparently I missed the arch-dependent bit of this. Fixed.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
tests/subsys/logging/log_core: increase log buffer size

The test needs to cram at least 8 messages into the log buffer, but
with 64-bit pointers only 5 fit, so make the buffer a bit bigger.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

gnuless commented:
This is ready for review - the CI failure appears to be spurious (unrelated RISCV32 board).

gnuless commented:
@andrewboie Would appreciate your review and feedback when you have the opportunity.

gnuless commented:
The only "ugly" aspect of this platform at present is the treatment of SSE: for the moment, the safe and easy method is used -- saving/restoring SSE state for every preemptive context switch.  Reasons for avoiding optimizations right now:

1. Unlike 32-bit mode, it is not enough to simply avoid floating point math to ensure the FP state is undisturbed -- SSE has more uses than that -- so we can't rely on a user-selected thread flag like `K_SSE_REGS` to speed context switches. 

2.Telling GCC to not use SSE at all breaks floating point support globally, because the 64-bit ABI relies on register calling conventions, including floating-point.  So functions which accept floating-point arguments or return a floating-point values cause GCC to bomb: there's just no (standard) mechanism for this.

3. Lazy FP switching has some security implications (which may or may not be relevant).  Also, with SMP support planned, this becomes a great deal more complex: think of the case where your FP state is still live on a *different* processor than the one you're scheduled on.

I'm not saying these problems are intractable, but they definitely require a great deal more thought than is necessary to get this platform support added.  If/when this is merged I'll open an issue to discuss and fix,

gnuless commented:
Force-push to nudge CI.

andyross commented:
> I don't think they plan to. I think the concern is that -- in the absence of a more developed firmware environment like UEFI

Long mode requires paging to be enabled.  Supporting this would mean that the hardware emulator would have to be in the business not just of jiggering the machine state to load a single binary image after initialization (itself kinda complicated if you think about it -- how does it know firmware is finished initializing?), but of allocating and populating page tables in a way that can be cleanly undone by whatever runs.  That's just too much work to be worth it vs. just running a UEFI firmware and letting it handle the complexity.

gnuless commented:
Force-push due to rebase (merge conflict).

andyross commented on arch/x86/core/intel64/locore.S:
```diff
@@ -0,0 +1,599 @@
+__resume:
+	movq $_kernel, %rdi
+	movq _kernel_offset_to_ready_q_cache(%rdi), %rsi
+	movq %rsi, _kernel_offset_to_current(%rdi)
```
So... this scheme doesn't work with SMP.  There is no ready_q_cache variable possible there and you need to call into the scheduler to fetch the next thread both here and in interrupt exit.  Is there a reason you dropped the z_arch_switch() implementation that already gave you that?

gnuless replied:
> Finally got some time to look at this closely, or at least at the core layer. Lots of notes, mostly of the form "my earlier stuff was better, please bring those bits back".

This should really be viewed as a forward-port of the x86 32-bit arch to 64-bit, not as some residual from your xuk/x86_64 work.  I didn't base this work on yours.  I didn't even really look at it once I decided it wasn't the right path forward for my use cases.

> Most of them are things that will only matter to microbenchmarks, or people trying to maintain or evolve the code long term, I guess (I mean, I honestly kind of hate your context switch code, but it clearly looks like it works). But the __swap() vs. z_arch_switch() thing is something you're going to need to fix for sure before SMP lands (er, lands again, I guess).

The requirement for `z_arch_switch()` with the SMP scheduler is an artifact of your implementation of said scheduler, not a reflection of some inherent problem with `__swap()` primitive on SMP.  I quite deliberately constructed this code with the intent that it be usable in multi-processor systems.  More on that later.

> A last thing that doesn't really fit in a line comment is that the xuk layer had a z_arch_printk_char_out() handler so that printk() would work (to either the serial port or the VGA console) from the very first 64 bit instruction. That really would be nice to have back.

The x86 32-bit doesn't have this, so neither does the 64-bit version.  Again, I didn't base my work on xuk/x86_64, I didn't take anything away that I can "give back".

andyross commented on arch/x86/core/intel64/locore.S:
```diff
@@ -0,0 +1,599 @@
+	pushq %rdx
+	pushq %rcx
+	pushq %rbx
+	pushq %rax
```
The older implementation was doing a trick where it would dump only the callee-saved registers here before entering the handler, saving a few cycles.  Any reason that got dropped?

gnuless replied:
You're right.  In the nested IRQ case, I am saving too much state.  I will fix that.

gnuless replied:
Fixed.

andyross commented on arch/x86/core/intel64/locore.S:
```diff
@@ -0,0 +1,599 @@
+	pushq %r12
+	pushq %r13
+	pushq %r14
+	pushq %r15
```
This duplicates the exception save code above in except(), more or less needlessly.   The original implementation had a single handler (at the cost of an instruction or three to handle the error code on some exceptions).

gnuless replied:
The exception path may look quite similar, but it's deliberately separated from the IRQ path, because they're different things.  At present, exceptions are all treated as fatal, and all the pushing there is really just to create an exception frame for the handler to dump.  So they're superficially nearly identical, but they're conceptually different, and will diverge over time, so they're uncoupled.

andyross commented on arch/x86/core/intel64/locore.S:
```diff
@@ -0,0 +1,599 @@
+	movq %rax, _thread_offset_to_rflags(%rsi)
+	popq %rax /* RSP */
+	movq %rax, _thread_offset_to_rsp(%rsi)
+	popq %rax /* SS: discard */
```
This is a **third** implementation of context save, with different semantics because you're saving to the thread struct instead of the stack (the original code always put the context on whichever stack it was on).   Again, I don't see the value here vs. having just one block for entry code.  Every new complexity that gets added needs to be fixed in triplicate.

This also has a non-trivial performance cost, as those indirected mov instructions encode using 2-4x as many bytes as a simple push, and the copy of the exception frame at the end is simple waste vs. just leaving it on the stack to begin with.

gnuless replied:
> This is a **third** implementation of context save, with different semantics because you're saving to the thread struct instead of the stack (the original code always put the context on whichever stack it was on). Again, I don't see the value here vs. having just one block for entry code. Every new complexity that gets added needs to be fixed in triplicate.

Remember, we're not `-fno-red-zone` so things are a bit more complicated than they might be on other platforms that can nest on one stack.

There are three context save scenarios, and they're all a bit different:

1. A deliberate __swap() out: this saves **partial** state to the **k_thread**.
2. An unnested IRQ: this saves **complete** state to the **k_thread**.
3. A nested IRQ: this saves **partial** state to the **stack**.

(As you pointed out, I'm saving complete state currently in case 3, but I will fix that.)

The reason case 2 and case 3 differ is because an unnested IRQ may result in thread preemption, and that thread may be resumed from a different CPU and thus we can't rely on any current-CPU state to be correct.  This is obviously not relevant *now*, because SMP doesn't work with `__swap()`, but that's a different discussion.

> This also has a non-trivial performance cost, as those indirected mov instructions encode using 2-4x as many bytes as a simple push

Actually, AMD recommends the MOV over the PUSH: the MOV is easier to parallelize because it has no hidden dependencies (read: RSP).  In general, there's not much of a correlation between the size of an instruction and its execution penalty on these large-core superscalar processors.  If we're writing for a "generic 64-bit target" it's often pointless to engage in these kinds of micro-optimization, because different micro-architectures will do better with different sequences.

> and the copy of the exception frame at the end is simple waste vs. just leaving it on the stack to begin with.

Can't leave it on the stack.  The thread might be preempted, and the interrupt stack is unique to the CPU, not the thread.

andyross commented on arch/x86/core/intel64/locore.S:
```diff
@@ -0,0 +1,599 @@
+	jz __resume
+
+irq_exit_nested:
+	fxrstor (%rsp)
```
Similarly, you have two separate assembly blocks for "switch to"

gnuless replied:
Yeah, the mechanism is necessarily different when leaving a nested IRQ than it is under other circumstances... so there are two separate assembly blocks.

andyross commented on arch/x86/core/intel64/locore.S:
```diff
@@ -0,0 +1,599 @@
+	.long 0
+	.fill 4088, 1, 0
+
+pdp:	.long 0x00000183		/* 0x183 = G, 1GB, R/W, P */
```
The original implementation unmapped the first 16k as a NULL guard.  Would be very useful to have that back.

gnuless replied:
I didn't take anything away, again, there is nothing to "give back".  The current behavior is in like with the 32-bit x86 port: there's no unmapped 0 page to catch NULL pointers.  When memory protection is added, we'll manipulate the page tables but this 2-page set of initial page tables will remain (and simply be modified at runtime as required).

andyross commented on arch/x86/core/intel64/locore.S:
```diff
@@ -0,0 +1,599 @@
+#include <arch/cpu.h>
+
+.section .locore,"ax"
+.code32
```
So... the existing x86_64 port has a 32 bit stub that runs as C code and can do stuff like set up early logging, interpret boot parameters (e.g. "am I startup code or a SMP CPU?"), parse a multiboot header (or UEFI if we need to run in a 32 bit UEFI context), set MSRs before mode switch maybe, etc...

The startup code below seems tight and sane, so it's not like it's a problem to do this all in assembly, but it seems like that was a fairly useful feature that got dropped.

gnuless replied:
If you can provide some technical justification for needing 32-bit C here, I'm all ears.  I don't see how any of the things you mention can't be done after we're in long mode.

andyross replied:
Doesn't this go the other way: what's the technical justifcation for throwing out the C boot stubs given that they're already implemented?  Why is this better than what's already commited?

gnuless replied:
There's no need for this intermediate 32-bit stage or the Byzantine build/link/image process you use to implement it.  You claim this "feature" is "useful" because it parses the multiboot header (by which I assume you mean boot info struct), but that doesn't even work:

```
        /* The multiboot memory map turns out not to be very useful.
         * The basic numbers logged here are only a subset of the true
         * memory map if it has holes or >4G memory, and the full map
         * passed in the second argument tends to live in low memory
         * and get easily clobbered by our own muckery.  If we care
         * about reading memory maps at runtime we probably want to be
         * using BIOS e820 like Linux does.
         */
```
Bizarrely, while you spell out the exact problem (that `xuk` clobbers the memory map before it's read) you seem to imply this is a failing of multiboot, rather than your code.  And your proposed solution to this problem is even worse: fall back into real mode and use the BIOS to read the E820 map?  That not only won't work on many systems (including our targets), it's just doubling down on a flawed design.

Do I really need to justify using my alternative, which, by your own admission, is "sane and tight", and most importantly, works?

andyross commented on arch/x86/core/intel64/locore.S:
```diff
@@ -0,0 +1,599 @@
+	movq %r12, _thread_offset_to_r12(%rsi)
+	movq %r13, _thread_offset_to_r13(%rsi)
+	movq %r14, _thread_offset_to_r14(%rsi)
+	movq %r15, _thread_offset_to_r15(%rsi)
```
By my count this makes four separate context save blocks.  The original code made do with just one, and z_arch_swich() was just a 6-instruction inline assembly block which used the embedded CALL instruction to populate the stack frame that would be used to restore.

Also this has the problem mentioned elsewhere that doing a stack save with MOV vs. PUSH involves trippling your icache footprint.

gnuless replied:
> By my count this makes four separate context save blocks. The original code made do with just one, and z_arch_swich() was just a 6-instruction inline assembly block which used the embedded CALL instruction to populate the stack frame that would be used to restore.

There are **three** context save (and restore) scenarios, not four.  (The exception "context save" is not a context save at all.)

Comparing this to the "original code" somehow implies that I derived this implementation from yours.  I did not.  This implementation is different and has different design constraints, and as a result I made different design choices.  Things are complicated here than your implementation because I don't use `-fno-red-zone` and have to use the IST mechanism.  You're comparing apples and oranges.

> Also this has the problem mentioned elsewhere that doing a stack save with MOV vs. PUSH involves trippling your icache footprint.

See previous comments.  The performance implications of MOV vs PUSH are not as cut-and-dried as you make them out to be, and I hardly think a few hundred more bytes in a .. what, 32K i-cache? .. is worth getting excited about.

gnuless replied:
And really, here, w/r/t icache: we're not even talking about a 'few' hundred more bytes.  Probably less than a hundred, total, across the entirely of the locore, with the MOVs vs. PUSHes. Even this could be trimmed down if the arch register struct weren't the last member of `k_thread`, though there appears to be (or at least, was at one time) a reason this member is at the end. But again, I don't the icache footprint is a worthy of much attention here.

andyross reviewed (commented):
Finally got some time to look at this closely, or at least at the core layer.   Lots of notes, mostly of the form "my earlier stuff was better, please bring those bits back".

Most of them are things that will only matter to microbenchmarks, or people trying to maintain or evolve the code long term, I guess (I mean, I honestly kind of hate your context switch code, but it clearly looks like it works).  But the __swap() vs. z_arch_switch() thing is something you're going to need to fix for sure before SMP lands (er, lands again, I guess).

A last thing that doesn't really fit in a line comment is that the xuk layer had a z_arch_printk_char_out() handler so that printk() would work (to either the serial port or the VGA console) from the very first 64 bit instruction.  That really would be nice to have back.

gnuless commented on include/arch/x86/arch.h:
```diff
@@ -16,6 +16,16 @@
+{
+	if ((key & 0x00000200U) != 0U) { /* 'IF' bit */
+		__asm__ volatile (
+			"sti;\n\t"
```
I didn't write this, it was existing x86 code.  I merely moved it.

gnuless commented on include/arch/x86/ia32/arch.h:
```diff
@@ -438,35 +387,21 @@ static ALWAYS_INLINE
+		: "=g" (key)
+		:
+		: "memory"
+		);
```
See above.  You're critiquing stuff that's been in the codebase for years.  Eventually I'll go back and clean this up, and indeed, I've done some cleanup along the way here, but that's quite obviously not the main thrust of this PR.

Commit:
arch/x86: optimize nested IRQ entry/exit

We don't need to save the ABI caller-save registers here, because
we don't preempt threads from nested IRQ contexts.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: clean up inline assembly format

Not sure why these were formatted the way they were, but I've removed
unnecessary line breaks, embedded control characters (actual embedded
tabs, not escape sequences) and escape sequences.

No functional change.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

andyross commented:
> This should really be viewed as a forward-port of the x86 32-bit arch to 64-bit, not as some residual from your xuk/x86_64 work. I didn't base this work on yours. I didn't even really look at it once I decided it wasn't the right path forward for my use cases.

That's... sort of a surprise.  I was pretty sure the intent here was to replace the existing work and remove it once done, right?  If so... shouldn't it at least have feature parity?  Believe me, no one is more a fan of "rewrite it and throw it away" than I am, but you're throwing away some really good stuff here...

Please go back and look at the older work and try to improve it.  We're trying to keep to one architecture for x86, right, not three?

gnuless commented:
Rebased to play nicely with @andrewboie's recent error-handling changes.

andyross commented:
> The requirement for z_arch_switch() with the SMP scheduler is an artifact of your implementation of said scheduler, not a reflection of some inherent problem with __swap() primitive on SMP.

No.  We need one architecture context switch API going forward.  If you want to replace switch, that's fine, but no rolling your own just to be different, please.

Just to clarify my position here, since some of the criticism was more serious than others.  The stuff this must have to get a +1 from me:

* Early boot char_out hook so you can printk() as the first line of z_cstart()
* z_arch_switch() based context switching, so it works with SMP unmodified, please don't argue this, that's the existing architecture.  If you want to fix that you need to fix it in the existing architectures (xtensa and x86_64 that already use it).
* Must have a qemu target and pass all tests that qemu_x86_64 does

andyross commented:
Oh, forgot:

* An unmapped NULL guard for the first few pages.  Optionally configurable, and I don't care that much about the size.  But routine code wants to be able to rely on this, even in kernel mode.

gnuless commented:
> Just to clarify my position here, since some of the criticism was more serious than others.

Nice attempt to walk back the "I think I'm smarter than I am" criticism.

> The stuff this must have to get a +1 from me:

Luckily for the project, I'm not particularly interested in your approval.

andrewboie commented:
Just now getting caught up with the discussion here. @gnuless Why was this PR closed and reopened in another PR? I reviewed the other one and didn't even see this discussion until later. Anyway, here goes...

> Early boot char_out hook so you can printk() as the first line of z_cstart()

@andyross I don't see this as necessary, or even particularly important. None of our other arches support this. There is not a lot of code that runs before cstart and it's straightforward to step through it in an attached GDB if something isn't working. I definitely didn't feel like I needed this when I wrote the Nios II port years ago. So NACK on this requirement.

> z_arch_switch() based context switching, so it works with SMP unmodified, please don't argue this, that's the existing architecture. If you want to fix that you need to fix it in the existing architectures (xtensa and x86_64 that already use it).

I do concur here. I don't see anything gained by having to special-case supporting swap-type switching in the core kernel SMP code. Let's just do it one way on all arches; if the design has problems we should take a holistic view of fixing them (although I'm not aware of any and if there are let's get some GH Issues opened please)

> Must have a qemu target and pass all tests that qemu_x86_64 does

This PR does add a QEMU target unless I'm missing something.

> The current behavior is in like with the 32-bit x86 port: there's no unmapped 0 page to catch NULL pointers.

Well, sorta. The QEMU targets anyway start memory at 4K, so that the NULL page is non-present. But this was implemented at the board/soc level definitions, it would be good to make this explicit and board/soc agnostic if we are starting memory at 0x0. I'm fine with punting on this part until we start bringing in memory protection and we have more fleshed out page tables, no need to do this in the initial commit.

>> Just to clarify my position here, since some of the criticism was more serious than others.
> Nice attempt to walk back the "I think I'm smarter than I am" criticism.
>> The stuff this must have to get a +1 from me:
> Luckily for the project, I'm not particularly interested in your approval.

Whoa whoa whoa no need for this discussion to escalate so much!

andrewboie commented:
> I do concur here. I don't see anything gained by having to special-case supporting swap-type switching in the core kernel SMP code. Let's just do it one way on all arches; if the design has problems we should take a holistic view of fixing them (although I'm not aware of any and if there are let's get some GH Issues opened please)

Just to clarify: we are gonna need to use switch when we implement SMP. But that doesn't need to be the initial PR. If the initial PR is uniprocessor and uses swap, I don't have any heartburn about it.
